src/utils/daos_hdlr.c

```diff
@@ -1535,6 +1536,758 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+			if (path_length >= MAX_FILENAME) 
+				fprintf(stderr, "Path length is too long.\n");
+
+		/* stat the source file */
+		struct stat st;
+		rc = daos_file_stat(daos_src_file, filename, &st);
+		if (rc)
+			fprintf(stderr, "Cannot stat path %s, %s\n",
+				d_name, strerror(errno));
+
+		if (S_ISREG(st.st_mode)) {
+			if (daos_src_file->type == DAOS && daos_dst_file->type == POSIX) {
+				daos_file_open(daos_src_file, filename, O_CREAT | O_RDWR, st.st_mode);
+				daos_file_open(daos_dst_file, dst_filename, O_CREAT | O_RDWR, st.st_mode);
+			} else {
+				daos_file_open(daos_src_file, filename, O_CREAT | O_RDWR, st.st_mode);
+				daos_file_open(daos_dst_file, filename, O_CREAT | O_RDWR, st.st_mode);
+			}
+
+			/* read from source file, then write to destination file */
+			uint64_t total_bytes = st.st_size;
+			uint64_t left_to_read = 0;
+			uint64_t buf_size = 64*1024*1024;
+			void *buf;
+			D_ALLOC(buf, buf_size * sizeof(char));
+				if (buf == NULL)
+					return ENOMEM;
+			while (left_to_read < total_bytes) {
+				left_to_read = total_bytes - left_to_read;
+				if (left_to_read > buf_size)
+					left_to_read = buf_size;
+				ssize_t bytes_read = daos_file_read(daos_src_file, filename,
+								    buf, left_to_read);
+				if (bytes_read < 0) {
+					fprintf(stderr, "read failed on %s\n", filename);
+				}
+				size_t bytes_to_write = (size_t) bytes_read;
+				ssize_t bytes_written;
+				if (daos_src_file->type == DAOS && daos_dst_file->type == POSIX) {
+					bytes_written = daos_file_write(daos_dst_file, dst_filename, buf, bytes_to_write);
+				} else {
+					bytes_written = daos_file_write(daos_dst_file, filename, buf, bytes_to_write);
+				}
+				if (bytes_written < 0) {
+					fprintf(stderr, "write failed on %s\n", filename);
+				}
+				left_to_read += (uint64_t) bytes_read;
+			}
+			if (buf != NULL)
+				D_FREE(buf);
+			daos_file_close(daos_src_file, filename);
+			daos_file_close(daos_dst_file, filename);
+		}
+
+		if (S_ISDIR(st.st_mode)) {
+			/* Check that the directory is not "d" or d's parent. */
+			if ((strcmp(d_name, "..") != 0) && (strcmp(d_name, ".") != 0)) {
+				char path[MAX_FILENAME];
+				char dpath[MAX_FILENAME];
+				path_length = snprintf(path, MAX_FILENAME, "%s", filename);
+				if (daos_src_file->type == DAOS && daos_dst_file->type == POSIX)
+					path_length = snprintf(dpath, MAX_FILENAME, "%s", dst_filename);
+				if (path_length >= MAX_FILENAME) 
+					fprintf(stderr, "Path length is too long.\n");
+				/* TODO create dst with write permissions, then after copying
+		 		 * make sure to change permissions to match the source */
+				if (daos_src_file->type == DAOS && daos_dst_file->type == POSIX) {
+					rc = daos_file_mkdir(daos_dst_file, dpath, st.st_mode);
+				} else {
+					rc = daos_file_mkdir(daos_dst_file, path, st.st_mode);
+				}
+				/* Recursively call "fs_copy" with the new path. */
+				rc = fs_copy(daos_src_file, daos_dst_file, path, fs_dst_prefix);
+			}
+		}
```

Comment:
@daltonbohning do you mean return ENOTSUP? if it is not a DIR or regular file in these situations?

Reply:
@dsikich Yes. instead of silently skipping over anything that isn't a file or directory, I think we should return an error. Or, at the very least, log a message/error so there is a record of it